_docs/rfc/scaffolding_new_modules.md

```diff
@@ -0,0 +1,402 @@
+  If it is a relative path, will assume it is relative to the `infrastructure_templates_repo`. This can be used to
+  override the location of the template, as an alternative to the convention followed in `infrastructure-templates`, or
+  even source from a completely different repository.
+- `new_region_template_source`, `new_account_template_source`, `new_wrapper_module_template_source`: Similar to
+  `new_environment_template_source`, but for each of the other scenarios that `terragrunt new` supports.
+
+### Template repository structure
+
+The `new` command will assume the following template repository structure:
+
+```
+infrastructure-templates-example
+├── CONTRIBUTING.md
+├── LICENSE.txt
+├── README.md
+├── templates
+│   ├── account/
+│   ├── component/
+│   ├── environment/
+│   ├── region/
+│   └── wrapper-module/
+└── test
+```
+
+Where `templates` and `test` are directories that contain the template folders and corresponding tests, respectively.
+
+By convention, templates in the `infrastructure-templates` repo with the following names will automatically be used for
+each of the `new` command use cases:
+
+- `component` => `terragrunt new component`
+- `environment` => `terragrunt new environment`
+- `region` => `terragrunt new region`
+- `account` => `terragrunt new account`
+- `wrapper-module` => `terragrunt new wrapper-module`
+
+Note that the templates folder is not restricted to just these 4 templates: you are free to define as many or as little
+templates as you need. Additional templates that do not fit the convention can be invoked using `terragrunt new
+unicorn`.
+
+### Template engine
+
+Templates are written and rendered using the default Go Template engine. While there are many other template engines to
+consider, given that the templates are rendered using Terragrunt as a frontend, it makes the most sense to use the built
+in templating engine of Go.
+
+**Pros**:
+
+- Go templating is builtin to Go, and thus many Go tools use this (e.g `hugo`, `helm`). This means that it will be
+  familiar with the Go community.
+- There are lots of utilities we can use for improving the experience of working with Go Templates, such as
+  [Sprig](http://masterminds.github.io/sprig/).
+- Does not depend on external tools: can be builtin to `terragrunt`, and thus is portable.
+
+**Cons**:
+
+- Go templating is not necessary a good language to work with. You lose editor support, and some of the constructs are
+  not intuitive (E.g whitespace chomping).
+- Requires some work to embed into `terragrunt` so that arbitrary templates can be invoked.
+- We need to invent a DSL for handling and declaring template inputs.
+
+### Template tests
+
+Template tests can be invoked using the `shell` module in `terratest` that can be used to call `terragrunt new` using
+these templates with arbitrary input values. You can then apply the generated template to test the generated code
+against various variable inputs.
```

Comment:
Not sure I follow this?

Reply:
To test the templates that you write, you use terratest to invoke the template by calling `terragrunt new` (using `shell`) and then run the terraform/terragrunt code.